storm_tests/generic/circular_queue_test.c

```diff
@@ -0,0 +1,151 @@
+    assert_false(circular_queue_enqueue(queue, message, MESSAGE_SIZE));
+}
+
+void test_circular_queue_enqueue_before_tail(void **state)
+{
+    // Arrange
+    circular_queue_test_type *circular_queue_test = (circular_queue_test_type *) *state;
+    circular_queue_type *queue = circular_queue_test->queue;
+    void *message = circular_queue_test->message;
+
+    // Act & Assert
+    //
+    // Like in the test above, this should succeed.
+    for (int i = 0; i < 630; i++)
+    {
+        assert_true(circular_queue_enqueue(queue, message, MESSAGE_SIZE));
+    }
+
+    void *dequeued_message = circular_queue_dequeue(queue);
+    assert_ptr_not_equal(dequeued_message, NULL);
+
+    // There should now be room before the tail, like this:
+    //
+    //   +-----------------+------+------+------+------+-----+--------+
+    //   | free tail space | msg2 | msg3 | msg4 | msg5 | ... | msg630 |
+    //   +-----------------+------+------+------+------+-----+--------+
+    //   ^                 ^
+    // head              tail
+
+    assert_true(circular_queue_enqueue(queue, message, MESSAGE_SIZE));
+}
+
+void test_circular_queue_enqueue_between_head_and_tail(void **state)
+{
+    // Arrange
+    circular_queue_test_type *circular_queue_test = (circular_queue_test_type *) *state;
+    circular_queue_type *queue = circular_queue_test->queue;
+    void *message = circular_queue_test->message;
+
+    // Act & Assert
+    //
+    // Like in the test above, this should succeed.
+    for (int i = 0; i < 630; i++)
+    {
+        assert_true(circular_queue_enqueue(queue, message, MESSAGE_SIZE));
+    }
+
+    void *dequeued_message = circular_queue_dequeue(queue);
+    assert_ptr_not_equal(dequeued_message, NULL);
+
+    dequeued_message = circular_queue_dequeue(queue);
+    assert_ptr_not_equal(dequeued_message, NULL);
+
+    dequeued_message = circular_queue_dequeue(queue);
+    assert_ptr_not_equal(dequeued_message, NULL);
+
+    // There should now be room before the tail, like this:
+    //
+    //   +-----------------+------+------+------+------+-----+--------+
+    //   | free tail space | msg4 | msg5 | msg6 | msg7 | ... | msg630 |
+    //   +-----------------+------+------+------+------+-----+--------+
+    //   ^                 ^
+    // head              tail
+    //
+    // The point of this test is to ensure that the free tail space can be used up in full; the
+    // test_circular_queue_enqueue_before_tail() only attempts to enqueue a single message into
+    // the free tail space while this test enqueues multiple ones. So this can be considered a
+    // superset of test_circular_queue_enqueue_before_tail()
+
+    assert_true(circular_queue_enqueue(queue, message, MESSAGE_SIZE));
+    assert_true(circular_queue_enqueue(queue, message, MESSAGE_SIZE));
+    assert_true(circular_queue_enqueue(queue, message, MESSAGE_SIZE));
+
+    // The queue is now expected to be full.
+    assert_false(circular_queue_enqueue(queue, message, MESSAGE_SIZE));
+}
```

Comment:
This is one of the weaker parts of this PR:

- The number of tests are not covering all aspects of the enqueuing/dequeueing, _especially_ not when trying to enqueue messages close to the available maximum (i.e.; 100 byte left in the middle of the queue, enqueuing 99, 100, 101 byte-messages and ensuring that the right thing happens)

- The messages that got added late in the `circular_queue.c` file (`circular_queue_peek`, `circular_queue_get_maximum_enqueue_size`) don't have any tests. This is simply because they got added at the point of integrating the circular queue code into the `mailbox_send` and `mailbox_receive`, after the circular queue main mechanisms were already in place. I was too lazy/focused at that time to add tests for the new methods (and they are pretty much "selective copy paste" of the dequeue/enqueue code anyway, with most of the logic removed.)

   Put it like this: it sure wouldn't hurt and it wouldn't even be hard to add some tests for these methods, but my time is limited so I will not do it right now. If someone wants to go ahead and add a PR for this, by all means - be my guest.